numpyro/distributions/discrete.py

```diff
@@ -620,6 +622,16 @@ def sample(self, key, sample_shape=()):
     def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
+        if self.is_sparse and value.shape:
+            rate, value = promote_shapes(self.rate, value)
+            nonzero = value > 0
+            sparse_value = value[nonzero]
+            sparse_rate = rate[nonzero]
```

Comment:
Unfortunately, this won't work (the shape of sparse_value is dynamic). I think you'll need to use `vmap(cond)` here. Something like
```python
shape = jax.lax.broadcast_shape(self.batch_shape, jnp.shape(value))
rate = jnp.broadcast_to(self.rate, shape).reshape(-1)
value = jnp.broadcast_to(value, shape).reshape(-1)
log_prob = jax.vmap(lambda r, v: jax.lax.cond(
    v > 0, lambda _: jnp.log(r) * v - gammaln(v + 1) - r, lambda _: -r, operand=None))(rate, value)
return log_prob.reshape(shape)
```

Reply:
Hmm, I worry this will be expensive.  In @phylyc's application (and in my applications in the past) this is used as a likelihood with static data `value`.  Would it be possible to implement this assuming `value` and `nonzero` are static constants?  I think a highly-optimized limited case would actually be most useful.

Reply:
Yup, it is possible then. We can add a small check here for that. Let me do a bit of benchmarking...

Reply:
@fehiepsi can you recommend any reading to help me understand when `jax.vmap(... jax.lax.cond(...))` is needed versus when `jax.obs.index_add(..)` will suffice?

Reply:
I think [this section](https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html#to-jit-or-not-to-jit) explains why `y[x>0]` does not work when `x` is a [tracer](https://jax.readthedocs.io/en/latest/glossary.html#term-Tracer). I benchmarked and found that `vmap(cond(...))` does not give any improvement for sparse arrays (actually, it is even slower than the non-sparse version). 

```python
import jax
import jax.numpy as jnp
from jax.scipy.special import gammaln
import numpyro
import numpy as np

numpyro.set_platform("cpu")


@jax.jit
def fn(rate, value):
    return (jnp.log(rate) * value) - gammaln(value + 1) - rate


@jax.jit
def fn_index_add(rate):
    nonzero = value > 0
    sparse_value = value[nonzero]
    sparse_rate = rate[nonzero]
    return jax.ops.index_add(
        -rate, nonzero,
        jnp.log(sparse_rate) * sparse_value - gammaln(sparse_value + 1))


@jax.jit
def fn_vmap_cond(rate, value):
    return jax.vmap(lambda r, v: jax.lax.cond(
        v > 0, lambda _: jnp.log(r) * v - gammaln(v + 1) - r,
        lambda _: -r, operand=None))(rate, value)

N = int(1e7)
rate = jnp.exp(jax.random.normal(jax.random.PRNGKey(0), (N,)))
value = np.zeros(N, dtype=np.int32)
value[:10] = np.arange(1, 11, dtype=np.int32)
x = fn(rate, value)
x1 = fn_index_add(rate)
x2 = fn_vmap_cond(rate, value)
np.testing.assert_allclose(x, x1, atol=1e-5)
np.testing.assert_allclose(x, x2, atol=1e-5)

%timeit x = fn(rate, value).copy()
%timeit x1 = fn_index_add(rate).copy()
%timeit x2 = fn_vmap_cond(rate, value).copy()
```

So in this case, we can do
```python
value = jax.device_get(value)
if self.is_sparse and not isinstance(value, jax.core.Tracer):
    ...
```

Reply:
Thanks for the help @fehiepsi !  It looks like this is working now.